### ISSUE TITLE
pyarrow can I use write_dataset  to write CSV as gzip compressed ?

### ISSUE THREAD

**Issue by eminrollic** [apache/arrow#34085]
### Describe the usage question you have. Please include as many useful details as  possible.

Hi,

I write a dataset to S3 with partitioned on a column as CSV files. But my CSV files should be gzip compressed, I couldn't find anything on docs to do that. Only way it seems, I should write with compressed outputstream however it's not one file.

This is the code I'm using;

```
ds.write_dataset(data, write_path,
                 format='csv',
                 partitioning=["p_col"],
                 partitioning_flavor="hive",
                 existing_data_behavior='delete_matching')
```

Is there anyway to achieve this to be output gzip files ?

If it's not available, I think it's a good enhancement to consider. 

### Component(s)

Format, Python

**Comment by westonpace** [comment]
No.  It is not possible to do today.  I agree it would be nice to support.  In the reader we have the concept of an "arbitrary input stream transform" (that transform can be decompression).  I think the writer could have a similar concept (with a compression transform).  It would only be useful for the text formats (csv, json).

CC @benibus

**Comment by westonpace** [comment]
`arrow::dataset::CsvFragmentScanOptions::stream_transform_func` is the read equivalent I am thinking of.  Seems we could do a similar option in `arrow::dataset::CsvFileWriteOptions`

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
